lib/req/steps.ex

```diff
@@ -752,29 +756,41 @@ defmodule Req.Steps do
   @doc step: :response
   def follow_redirects(request_response)
 
-  def follow_redirects({request, %{status: status} = response})
-      when status in [301, 302, 303, 307, 308] do
-    {_, location} = List.keyfind(response.headers, "location", 0)
-    Logger.debug(["follow_redirects: redirecting to ", location])
-
-    location_trusted = Map.get(request.options, :location_trusted)
-    location_url = URI.parse(location)
+  def follow_redirects({request, %{status: status} = response}) when status in 301..308 do
+    max_redirects = Map.get(request.options, :max_redirects, 2)
+    redirect_count = Req.Request.get_private(request, :req_redirect_count, 0)
 
-    request =
-      request
-      |> remove_params()
-      |> remove_credentials_if_untrusted(location_trusted, location_url)
-      |> put_redirect_request_method()
-      |> put_redirect_location(location_url)
+    if redirect_count < max_redirects do
+      request =
+        request
+        |> build_redirect_request(response)
+        |> Req.Request.put_private(:req_redirect_count, redirect_count + 1)
 
-    {_, result} = Req.Request.run(request)
-    {Req.Request.halt(request), result}
+      {_, result} = Req.Request.run(request)
+      {request, result}
+    else
```

Comment:
let's follow what curl does: if we have exceeded max redirects, let's raise an error: `"too many redirects (#{max_redirects})"`